File: impl/packages/core/src/jws.js
// Compact JWS (EdDSA/Ed25519) over detached JSON payloads, using node:crypto only.
import { generateKeyPairSync, sign as edSign, verify as edVerify, createPublicKey, createPrivateKey } from 'node:crypto';

const b64u = (buf) => Buffer.from(buf).toString('base64url');
const fromB64u = (s) => Buffer.from(s, 'base64url');

/** Deterministic JSON: object keys sorted recursively, arrays kept in order. */
export function stableStringify(value) {
  if (Array.isArray(value)) return '[' + value.map(stableStringify).join(',') + ']';
  if (value && typeof value === 'object') {
    return '{' + Object.keys(value).sort().map((k) => JSON.stringify(k) + ':' + stableStringify(value[k])).join(',') + '}';
  }
  return JSON.stringify(value);
}

export function generateKeypair() {
  const { publicKey, privateKey } = generateKeyPairSync('ed25519');
  return {
    publicKey,
    privateKey,
    publicJwk: publicKey.export({ format: 'jwk' }),
    privatePem: privateKey.export({ type: 'pkcs8', format: 'pem' }),
  };
}

export function keyFromPem(pem) {
  return createPrivateKey(pem);
}

/** Rebuild the full keypair shape from a stored PKCS8 PEM (persistent identities). */
export function keypairFromPem(pem) {
  const privateKey = createPrivateKey(pem);
  const publicKey = createPublicKey(privateKey);
  return { publicKey, privateKey, publicJwk: publicKey.export({ format: 'jwk' }), privatePem: pem };
}

export function keyFromJwk(jwk) {
  return createPublicKey({ key: jwk, format: 'jwk' });
}

/** Sign a JSON payload; returns compact JWS string header.payload.signature. */
export function signJson(payload, privateKey, kid = 'onp-1') {
  const header = b64u(JSON.stringify({ alg: 'EdDSA', typ: 'JWT', kid }));
  const body = b64u(stableStringify(payload));
  const sig = edSign(null, Buffer.from(`${header}.${body}`), privateKey);
  return `${header}.${body}.${b64u(sig)}`;
}

/** Verify a compact JWS against a public key; returns the decoded payload or throws. */
export function verifyJws(jws, publicKey) {
  const parts = jws.split('.');
  if (parts.length !== 3) throw new Error('malformed JWS');
  const [header, body, sig] = parts;
  const ok = edVerify(null, Buffer.from(`${header}.${body}`), publicKey, fromB64u(sig));
  if (!ok) throw new Error('JWS signature invalid');
  return JSON.parse(fromB64u(body).toString('utf8'));
}

export function decodeJwsHeader(jws) {
  return JSON.parse(fromB64u(jws.split('.')[0]).toString('utf8'));
}

File: node-kit/onp_node/jws.py
"""Compact JWS (EdDSA/Ed25519) byte-compatible with the JS reference implementation.

The registry verifies card signatures by re-serializing the received card with its own
stable stringify (sorted keys, compact separators, raw unicode) and checking the JWS
over those exact bytes — so `stable_stringify` here MUST byte-match the JS version
for any value that appears in a card.
"""
import base64
import json

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey


def b64u(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def from_b64u(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def _normalize(v):
    # JS JSON.stringify(0.0) === "0": integral floats must serialize as ints to byte-match.
    if isinstance(v, bool):
        return v
    if isinstance(v, float) and v.is_integer():
        return int(v)
    if isinstance(v, dict):
        return {k: _normalize(x) for k, x in v.items()}
    if isinstance(v, list):
        return [_normalize(x) for x in v]
    return v


def stable_stringify(value) -> str:
    # Matches JS: recursive key sort, arrays in order, compact separators, no \u escapes.
    return json.dumps(_normalize(value), sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def generate_private_key() -> Ed25519PrivateKey:
    return Ed25519PrivateKey.generate()


def private_key_to_pem(key: Ed25519PrivateKey) -> str:
    return key.private_bytes(
        serialization.Encoding.PEM,
        serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption(),
    ).decode()


def private_key_from_pem(pem: str) -> Ed25519PrivateKey:
    return serialization.load_pem_private_key(pem.encode(), password=None)


def public_jwk(key: Ed25519PrivateKey, kid: str = "onp-1") -> dict:
    raw = key.public_key().public_bytes(
        serialization.Encoding.Raw, serialization.PublicFormat.Raw
    )
    return {"kty": "OKP", "crv": "Ed25519", "x": b64u(raw), "kid": kid}


def sign_json(payload, key: Ed25519PrivateKey, kid: str = "onp-1") -> str:
    header = b64u(json.dumps({"alg": "EdDSA", "typ": "JWT", "kid": kid},
                             separators=(",", ":")).encode())
    body = b64u(stable_stringify(payload).encode())
    sig = key.sign(f"{header}.{body}".encode())
    return f"{header}.{body}.{b64u(sig)}"


def verify_jws(jws: str, jwk: dict):
    header, body, sig = jws.split(".")
    pub = Ed25519PublicKey.from_public_bytes(from_b64u(jwk["x"]))
    pub.verify(from_b64u(sig), f"{header}.{body}".encode())  # raises on mismatch
    return json.loads(from_b64u(body))


def sign_card(card: dict, key: Ed25519PrivateKey, kid: str = "onp-1") -> dict:
    payload = {k: v for k, v in card.items() if k != "signatures"}
    jws = sign_json(payload, key, kid)
    protected, _, signature = jws.split(".")
    return {**card, "signatures": [{"protected": protected, "signature": signature, "kid": kid}]}
